### DISCUSSION TITLE
Why is WGSL verbose?

### DISCUSSION THREAD

**Discussion by Jerrody** [gfx-rs/wgpu#2354] [Q&A]
Why am I asking about it? Let's look at this example of code:
```
struct VertexInput {
    [[location(0)]] position: vec3<f32>;
    [[location(1)]] color: vec3<f32>;
};

struct VertexOutput {
    [[builtin(position)]] clip_position: vec4<f32>;
    [[location(0)]] color: vec3<f32>;
};

[[stage(vertex)]]
fn vs_main(
    model: VertexInput,
) -> VertexOutput {
    var out: VertexOutput;
    out.clip_position = vec4<f32>(model.position, 1.0);
    out.color = model.color;

    return out;
}

[[stage(fragment)]]
fn fs_main(in: VertexOutput) -> [[location(0)]] vec4<f32> {
    return vec4<f32>(in.color, 1.0);
}
```

This is an example of code from [WGPU Learn](https://sotrh.github.io/learn-wgpu/beginner/tutorial4-buffer/#so-what-do-i-do-with-it). So, what do we see? What's wrong mate?
I will do a list of what I saw verbosity:
1) `[[location(0)]]`, why can't we just write `[location(0)]? And specify a few attributes with a comma? Also the same with everything that has `[[<...>]]`.
2) This is my preferences, but why can't we use of Rust's style return, like `vec4<f32>(in.color, 1.0)` instead of `    return vec4<f32>(in.color, 1.0);`.
3) Input parameters, everything is alright, but maybe for developer doesn't need to write a comma for the last parameter?
4) About structs, maybe don't need to write `;` after `}`. `}` tells that this is the end of defining a struct.

So, this is my opinion, but this issue is directed to the more clear and less verbose syntax of WGSL that I so much like.

**Comment by memoryruins** [comment]
`wgpu` does not define the specifications of WebGPU and WGSL.
[WGSL's spec](https://gpuweb.github.io/gpuweb/wgsl/) suggests <https://github.com/gpuweb/gpuweb/issues/> for feedback.

**Comment by kvark** [comment]
> why can't we just write [location(0)]

https://github.com/gpuweb/gpuweb/issues/2243

> And specify a few attributes with a comma?

You actually can! Example: `[[group(0), binding(1)]]`

> but why can't we use of Rust's style return

I'd love that as well, but WGSL has way less support for expressions than Rust, and at this point the `return` is needed.

> Input parameters, everything is alright, but maybe for developer doesn't need to write a comma for the last parameter?

They don't have to write comma after the last parameter. It's just consistent with Rust to do so (e.g. clippy may yell at you).

> About structs, maybe don't need to write ; after }. }

Was mentioned in https://github.com/gpuweb/gpuweb/issues/1466 and even supported, but didn't go further (separate issue is needed).
